File: Python/ilinkai_weixin/media/media_upload_service.py
"""媒体上传服务"""
import hashlib
import logging
import os
from dataclasses import dataclass

from ..cdn.aes_ecb_crypto import AesEcbCrypto
from ..cdn.cdn_client import CdnClient
from ..models import GetUploadUrlRequest, UploadMediaType

logger = logging.getLogger(__name__)


@dataclass
class UploadedFileInfo:
    file_key: str = ""
    download_encrypted_query_param: str = ""
    aes_key: str = ""
    file_size: int = 0
    file_size_ciphertext: int = 0


class MediaUploadService:
    def __init__(self, api_client, cdn_base_url: str):
        self._api_client = api_client
        self._cdn_base_url = cdn_base_url
        self._cdn_client = CdnClient()

    def upload_file(self, file_path: str, to_user_id: str, media_type: int) -> UploadedFileInfo:
        with open(file_path, 'rb') as f:
            plaintext = f.read()
        raw_size = len(plaintext)
        raw_file_md5 = hashlib.md5(plaintext).hexdigest()
        file_size = AesEcbCrypto.get_padded_size(raw_size)
        file_key = os.urandom(16).hex()
        aes_key = AesEcbCrypto.generate_key()

        logger.debug("UploadFile: file=%s rawSize=%d fileSize=%d md5=%s fileKey=%s",
                      file_path, raw_size, file_size, raw_file_md5, file_key)

        request = GetUploadUrlRequest(
            file_key=file_key, media_type=media_type, to_user_id=to_user_id,
            raw_size=raw_size, raw_file_md5=raw_file_md5, file_size=file_size,
            no_need_thumb=True, aes_key=aes_key.hex(),
        )
        upload_url_resp = self._api_client.get_upload_url(request)
        upload_param = upload_url_resp.upload_param
        if not upload_param:
            raise RuntimeError("getUploadUrl returned no upload_param")

        download_param = self._cdn_client.upload_buffer(
            plaintext, upload_param, file_key, self._cdn_base_url, aes_key,
            f"UploadFile[fileKey={file_key}]")

        return UploadedFileInfo(
            file_key=file_key, download_encrypted_query_param=download_param,
            aes_key=aes_key.hex(), file_size=raw_size, file_size_ciphertext=file_size,
        )

    def upload_image(self, file_path: str, to_user_id: str) -> UploadedFileInfo:
        return self.upload_file(file_path, to_user_id, UploadMediaType.IMAGE)

    def upload_video(self, file_path: str, to_user_id: str) -> UploadedFileInfo:
        return self.upload_file(file_path, to_user_id, UploadMediaType.VIDEO)

    def upload_file_attachment(self, file_path: str, to_user_id: str) -> UploadedFileInfo:
        return self.upload_file(file_path, to_user_id, UploadMediaType.FILE)

File: Java/src/main/java/com/ilinkai/weixin/media/MediaUploadService.java
package com.ilinkai.weixin.media;

import com.ilinkai.weixin.WeixinApiClient;
import com.ilinkai.weixin.WeixinApiException;
import com.ilinkai.weixin.cdn.AesEcbCrypto;
import com.ilinkai.weixin.cdn.CdnClient;
import com.ilinkai.weixin.cdn.CdnException;
import com.ilinkai.weixin.models.ApiModels;
import com.ilinkai.weixin.models.Enums;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.SecureRandom;

/** 媒体上传服务 */
public class MediaUploadService {
    private static final Logger logger = LoggerFactory.getLogger(MediaUploadService.class);

    private final WeixinApiClient apiClient;
    private final CdnClient cdnClient;
    private final String cdnBaseUrl;

    public MediaUploadService(WeixinApiClient apiClient, String cdnBaseUrl) {
        this.apiClient = apiClient;
        this.cdnBaseUrl = cdnBaseUrl;
        this.cdnClient = new CdnClient();
    }

    public UploadedFileInfo uploadFile(String filePath, String toUserId, int mediaType)
            throws IOException, WeixinApiException, CdnException {
        byte[] plaintext = Files.readAllBytes(Paths.get(filePath));
        int rawSize = plaintext.length;
        String rawFileMd5 = md5Hex(plaintext);
        int fileSize = AesEcbCrypto.getPaddedSize(rawSize);
        String fileKey = generateFileKey();
        byte[] aesKey = AesEcbCrypto.generateKey();

        logger.debug("UploadFile: file={} rawSize={} fileSize={} md5={} fileKey={}",
                filePath, rawSize, fileSize, rawFileMd5, fileKey);

        ApiModels.GetUploadUrlRequest request = new ApiModels.GetUploadUrlRequest();
        request.setFileKey(fileKey);
        request.setMediaType(mediaType);
        request.setToUserId(toUserId);
        request.setRawSize(rawSize);
        request.setRawFileMd5(rawFileMd5);
        request.setFileSize(fileSize);
        request.setNoNeedThumb(true);
        request.setAesKey(bytesToHex(aesKey));

        ApiModels.GetUploadUrlResponse uploadUrlResp = apiClient.getUploadUrl(request);
        String uploadParam = uploadUrlResp.getUploadParam();
        if (uploadParam == null || uploadParam.isEmpty()) {
            throw new IOException("getUploadUrl returned no upload_param");
        }

        String downloadParam = cdnClient.uploadBuffer(plaintext, uploadParam, fileKey, cdnBaseUrl, aesKey,
                "UploadFile[fileKey=" + fileKey + "]");

        UploadedFileInfo info = new UploadedFileInfo();
        info.fileKey = fileKey;
        info.downloadEncryptedQueryParam = downloadParam;
        info.aesKey = bytesToHex(aesKey);
        info.fileSize = rawSize;
        info.fileSizeCiphertext = fileSize;
        return info;
    }

    public UploadedFileInfo uploadImage(String filePath, String toUserId) throws IOException, WeixinApiException, CdnException {
        return uploadFile(filePath, toUserId, Enums.UploadMediaType.IMAGE);
    }

    public UploadedFileInfo uploadVideo(String filePath, String toUserId) throws IOException, WeixinApiException, CdnException {
        return uploadFile(filePath, toUserId, Enums.UploadMediaType.VIDEO);
    }

    public UploadedFileInfo uploadFileAttachment(String filePath, String toUserId) throws IOException, WeixinApiException, CdnException {
        return uploadFile(filePath, toUserId, Enums.UploadMediaType.FILE);
    }

    private static String generateFileKey() {
        byte[] bytes = new byte[16];
        new SecureRandom().nextBytes(bytes);
        return bytesToHex(bytes);
    }

    private static String md5Hex(byte[] data) {
        try {
            byte[] digest = MessageDigest.getInstance("MD5").digest(data);
            return bytesToHex(digest);
        } catch (Exception e) {
            throw new RuntimeException("MD5 not available", e);
        }
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) sb.append(String.format("%02x", b));
        return sb.toString();
    }

    public static class UploadedFileInfo {
        public String fileKey = "";
        public String downloadEncryptedQueryParam = "";
        public String aesKey = "";
        public int fileSize;
        public int fileSizeCiphertext;
    }
}
